packages/compass-shell/src/modules/runtime.js

```diff
@@ -38,22 +40,41 @@ export default function reducer(state = INITIAL_STATE, action) {
 
 function reduceSetupRuntime(state, action) {
   if (action.error || !action.dataService) {
-    return {error: action.error, dataService: null, runtime: null};
+    return { error: action.error, dataService: null, runtime: null };
   }
 
   if (state.dataService === action.dataService) {
     return state;
   }
 
-  const runtime = new ElectronRuntime(
-    CompassServiceProvider.fromDataService(action.dataService),
-    action.appRegistry
-  );
+  // const runtime = new ElectronRuntime(
+  //   CompassServiceProvider.fromDataService(action.dataService),
+  //   action.appRegistry
+  // );
+
+  // const origEval = runtime.evaluate;
+
+  // runtime.evaluate = async function _evaluate(...args) {
+  //   console.info("input:", ...args);
+  //   const result = await origEval.call(runtime, ...args);
+  //   console.info("output:", result);
+  //   return result;
+  // };
+
+  if (state.runtime) {
+    state.runtime.terminate();
+  }
+
+  const { url, options } = action.dataService.getConnectionParams();
+  const newRuntime = new WorkerRuntime(url, options);
+
+  // TODO: For testing purposes
+  window.restart = newRuntime.restart.bind(newRuntime);
```

Comment:
If you are running this branch locally, you can run a `while(true){}` in your shell and then manually call `restart` in the console, you'll see that the script just gracefully interrupts:

![Kapture 2020-12-10 at 12 16 33](https://user-images.githubusercontent.com/5036933/101765442-98c48280-3ae1-11eb-89b0-b433d8983eb9.gif)